### PR TITLE
Use standard `renderPagination` in list page for Jobs Dashboard

### PR DESCRIPTION
**Pull request by oclbdk**
Addresses #1198

# Test Plan

I tested locally by creating 65 job instances in a sample project and checking that the pagination renders as expected for various page and window sizes:

Screen Shot 2021-12-28 at 2 35 54 PM

## Cases

* No pagination is rendered when all items fit on a single page.
* Only the necessary number of pages are listed when `window size > total pages`.
* Only the specified number of items are rendered per page when selecting the "items per page" dropdown.
* Clicking on an arbitrary page number navigates to that page.
* Clicking on the "previous" and "next" arrows navigates to the associated pages.

# QueryBuilder / Limitations

There's still redundancy with `IHP.Pagination.ControllerFunctions.paginateWithOptions`.

I tried to use that function but couldn't figure out how to initialize a `QueryBuilder` with a `Text`-valued table name. This isn't possible, right? Since the static types from the application aren't accessible to the IHP library.

I'm wondering if it makes sense to duplicate the validation logic here? In particular, the logic to constrain the page size.

### CODE REVIEW

File: IHP/Job/Dashboard.hs [hunk 1]
```diff
@@ -176,11 +177,14 @@ instance JobsDashboard '[] where
...
-        render $ HtmlView $ renderBaseJobTablePaginated table jobs page numberOfPages
+        let table = param "tableName"
+            options = defaultPaginationOptions
+            page = paramOrDefault 1 "page"
+            pageSize = paramOrDefault (maxItems options) "maxItems"
+        totalItems <- totalRecordsForTable table
+        jobs <- queryBaseJobsFromTablePaginated table (page - 1) pageSize
+        let pagination = Pagination { currentPage = page, totalItems, pageSize, window = windowSize options }
+        render $ HtmlView $ renderBaseJobTablePaginated table jobs pagination
         render $ EmptyView
```
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:188]: The `render $ EmptyView` is unreachable. Can it be removed?

```suggestion
```

File: IHP/Job/Dashboard/Utils.hs [hunk 2]
```diff
@@ -7,10 +7,14 @@ import qualified Database.PostgreSQL.Simple.Types as PG
...
 numberOfPagesForTable table pageSize = do
-    (PG.Only totalRecords : _) <- sqlQuery
-        (PG.Query $ cs $ "SELECT COUNT(*) FROM " <> table)
-        ()
+    totalRecords <- totalRecordsForTable table
     pure $ case totalRecords `quotRem` pageSize of
         (pages, 0) -> pages
         (pages, _) -> pages + 1
 
+totalRecordsForTable :: (?modelContext::ModelContext) => Text -> IO Int
```
Comment (mpscholten) [inline IHP/Job/Dashboard/Utils.hs:15]:
```suggestion
totalRecordsForTable :: (?modelContext :: ModelContext) => Text -> IO Int
```

File: IHP/Job/Dashboard/Utils.hs [hunk 3]
```diff
@@ -7,10 +7,14 @@ import qualified Database.PostgreSQL.Simple.Types as PG
...
+totalRecordsForTable table = do
```
Comment (mpscholten) [inline IHP/Job/Dashboard/Utils.hs:16]: Next to `sqlQuery` we also have [`sqlQueryScalar`](https://ihp.digitallyinduced.com/api-docs/IHP-ModelSupport.html#v:sqlQueryScalar) which only expects a single scalar result value (e.g. a single Integer like here). Using this function we could write this a bit shorter:

```haskell
totalRecordsForTable table = sqlQueryScalar (PG.Query $ cs $ "SELECT COUNT(*) FROM " <> table) ()
```

**Comment by mpscholten** [comment]
Good job on the PR again :) Thanks! 👍 

> I tried to use that function but couldn't figure out how to initialize a QueryBuilder with a Text-valued table name. This isn't possible, right? Since the static types from the application aren't accessible to the IHP library.

Yes exactly. The QueryBuilder uses a lot of type magic to ensure that only valid columns etc. are used. This makes it very hard to use in these dynamic cases.

> I'm wondering if it makes sense to duplicate the validation logic here? In particular, the logic to constrain the page size.

The dashboard is designed for internal use only. So I'm fine if we just leave this out.

**Review by zacwood9 — APPROVED** [review]
Looks great! Thanks

**Review by mpscholten — APPROVED** [review]

**Pull request merged**